packages/gguf/scripts/generate-llm.ts

```diff
@@ -0,0 +1,194 @@
+		if (matched && !matched.groups?.name.match(/unknown/)) {
+			archList.push({
+				cppConst: matched.groups?.cppConst || "",
+				name: matched.groups?.name || "",
+				tsName: snakeToPascal(matched[1].replace("LLM_", "")),
+				tensorNames: [],
+				hparams: [],
+			});
+		}
+	}
+
+	/////////////////////////////////////
+	// extract map constant name to kv name
+	// for example: LLM_KV_ATTENTION_LAYERNORM_RMS_EPS ==> "%s.attention.layer_norm_rms_epsilon"
+	const constToKVName: { [cppConst: string]: string } = {};
+	const matchedKVList = cppSource.match(/LLM_KV_NAMES = (?<names>[^;]+)/)?.groups?.names.split("\n") || [];
+	if (matchedKVList.length === 0) {
+		throw new Error("LLM_KV_NAMES is empty");
+	}
+	for (const line of matchedKVList) {
+		const matched = line.match(/(LLM_KV_[A-Z0-9_]+)[,\s]+"([^"]+)"/);
+		if (matched) {
+			constToKVName[matched[1]] = matched[2];
+		}
+	}
+
+	/////////////////////////////////////
+	// extract list of tensor names based on architecture
+	// TODO: unused for now
+	const matchedTensorList = cppSource.match(/LLM_TENSOR_NAMES = (?<names>[^;]+)/)?.groups?.names.split("\n") || [];
+	if (matchedTensorList.length === 0) {
+		throw new Error("LLM_TENSOR_NAMES is empty");
+	}
+	let currCppConst = "";
+	for (const line of matchedTensorList) {
+		// check if current line has LLM_ARCH_*
+		const cppConst = (line.match(RE_ARCH_NAME) || [])[0];
+		if (cppConst) {
+			currCppConst = cppConst;
+			continue;
+		}
+		// check if current line has LLM_TENSOR_*
+		const tensorMatched = line.match(/LLM_TENSOR_[A-Z0-9_]+[,\s]+"(?<name>.+?)"/);
+		if (tensorMatched) {
+			const arch = archList.find((a) => a.cppConst === currCppConst);
+			if (arch) arch.tensorNames.push(tensorMatched.groups?.name || "");
+		}
+	}
+
+	/////////////////////////////////////
+	// extract list of hyper params based on architecture
+	let insideLoadHParamsFn = false;
+	currCppConst = "";
+	for (const line of cppSource.split("\n")) {
+		// check if current line is function llm_load_hparams()
+		if (line.startsWith("static void llm_load_hparams")) {
+			insideLoadHParamsFn = true;
+		}
+		if (!insideLoadHParamsFn) {
+			continue;
+		}
+		// check if current line has LLM_ARCH_*
+		const RE_CASE = new RegExp(`case (${RE_ARCH_NAME.source})`);
+		const cppConst = (line.match(RE_CASE) || [])[1];
+		if (cppConst) {
+			currCppConst = cppConst;
+			continue;
+		}
+		// check if current line has get_key(...)
+		const keyConst = (line.match(/LLM_KV_[A-Z0-9_]+/) || [])[0];
+		if (keyConst) {
+			const arch = archList.find((a) => a.cppConst === currCppConst);
+			if (arch) {
+				arch.hparams.push(keyConst);
+			}
+		}
+		// check if current line is end-of-function
+		if (line === "}") {
+			break;
+		}
+	}
+
+	/////////////////////////////////////
+	// write result to file
+	const content = [
+		'import type { ModelBase } from "./types";',
+		"// This file is auto-generated by generate-llm.ts",
```

Comment:
why not put this comment line on top of the file?

Reply:
I usually place import on the top then the comment. Just personal preference though, but you're right, it should be on top of the file.

Reply:
handled in https://github.com/huggingface/huggingface.js/commit/6962b5143a70631f5a2736b53ba1ddbdc8d609f0